From 086b1b7b09dce60ac25dabf4270ab769801b3502 Mon Sep 17 00:00:00 2001
From: Justin W Smith <103147162+justsmth@users.noreply.github.com>
Date: Mon, 6 Oct 2025 18:47:34 -0400
Subject: [PATCH] [FIPS 2024 - CHERRY_PICK] Offer P521 for signature_algorithms
 in client Hello (#2572) (#2731)
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Cherry-pick onto fips-2024-09-27. Original PR for main:
* #2572

I needed to resolve a merge conflict for the ssl_test.cc changes.

-------
### Issues:
n/a

### Description of changes:

This change adds ECDSA P521 to our list of supported certificate
verification algorithms in client Hello.

### Call-outs:
n/a

### Testing:

CI for automated tests. Manual test to confirm `signature_algorithms`
extension contents before/after:

```
.﻿﻿﻿/build/tool/bssl s_client -connect kms.us-east-1.amazonaws.com:443
```

Client Hello before:

<img width="1508" height="883" alt="Screenshot 2025-07-23 at 5 52 15 PM"
src="https://github.com/user-attachments/assets/def57f79-3113-44d4-865a-40bd22b47994"
/>

Client Hello after:

<img width="1508" height="882" alt="Screenshot 2025-07-23 at 5 47 51 PM"
src="https://github.com/user-attachments/assets/a7613af3-c2f7-4580-9964-5c86338c985a"
/>


---

By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache 2.0 license and the ISC license.

By submitting this pull request, I confirm that my contribution is made
under the terms of the Apache 2.0 license and the ISC license.

Co-authored-by: Will Childs-Klein <childw@amazon.com>
---
 ssl/extensions.cc                                |  1 +
 ssl/ssl_test.cc                                  | 12 ++++++------
 ssl/test/runner/runner.go                        |  4 ++--
 ssl/test/runner/ssl_transfer/test_case_names.txt |  2 ++
 4 files changed, 11 insertions(+), 8 deletions(-)

diff --git a/ssl/extensions.cc b/ssl/extensions.cc
index 8f134c09f..c7b32706b 100644
--- a/ssl/extensions.cc
+++ b/ssl/extensions.cc
@@ -392,6 +392,7 @@ static const uint16_t kVerifySignatureAlgorithms[] = {
     SSL_SIGN_RSA_PSS_RSAE_SHA384,
     SSL_SIGN_RSA_PKCS1_SHA384,
 
+    SSL_SIGN_ECDSA_SECP521R1_SHA512,
     SSL_SIGN_RSA_PSS_RSAE_SHA512,
     SSL_SIGN_RSA_PKCS1_SHA512,
 
diff --git a/ssl/ssl_test.cc b/ssl/ssl_test.cc
index 968f443b8..0cdec36ac 100644
--- a/ssl/ssl_test.cc
+++ b/ssl/ssl_test.cc
@@ -4735,18 +4735,18 @@ TEST(SSLTest, ClientHello) {
         0x0a, 0x00, 0x08, 0x00, 0x06, 0x00, 0x1d, 0x00, 0x17, 0x00, 0x18, 0x00,
         0x0b, 0x00, 0x02, 0x01, 0x00, 0x00, 0x23, 0x00, 0x00}},
       {TLS1_2_VERSION,
-       {0x16, 0x03, 0x01, 0x00, 0x86, 0x01, 0x00, 0x00, 0x82, 0x03, 0x03, 0x00,
+       {0x16, 0x03, 0x01, 0x00, 0x88, 0x01, 0x00, 0x00, 0x84, 0x03, 0x03, 0x00,
         0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
         0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
         0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x22, 0xcc, 0xa9,
         0xcc, 0xa8, 0xc0, 0x2b, 0xc0, 0x2f, 0xc0, 0x2c, 0xc0, 0x30, 0xc0, 0x09,
         0xc0, 0x13, 0xc0, 0x27, 0xc0, 0x0a, 0xc0, 0x14, 0xc0, 0x28, 0x00, 0x9c,
-        0x00, 0x9d, 0x00, 0x2f, 0x00, 0x3c, 0x00, 0x35, 0x01, 0x00, 0x00, 0x37,
+        0x00, 0x9d, 0x00, 0x2f, 0x00, 0x3c, 0x00, 0x35, 0x01, 0x00, 0x00, 0x39,
         0x00, 0x17, 0x00, 0x00, 0xff, 0x01, 0x00, 0x01, 0x00, 0x00, 0x0a, 0x00,
         0x08, 0x00, 0x06, 0x00, 0x1d, 0x00, 0x17, 0x00, 0x18, 0x00, 0x0b, 0x00,
-        0x02, 0x01, 0x00, 0x00, 0x23, 0x00, 0x00, 0x00, 0x0d, 0x00, 0x14, 0x00,
-        0x12, 0x04, 0x03, 0x08, 0x04, 0x04, 0x01, 0x05, 0x03, 0x08, 0x05, 0x05,
-        0x01, 0x08, 0x06, 0x06, 0x01, 0x02, 0x01}},
+        0x02, 0x01, 0x00, 0x00, 0x23, 0x00, 0x00, 0x00, 0x0d, 0x00, 0x16, 0x00,
+        0x14, 0x04, 0x03, 0x08, 0x04, 0x04, 0x01, 0x05, 0x03, 0x08, 0x05, 0x05,
+        0x01, 0x06, 0x03, 0x08, 0x06, 0x06, 0x01, 0x02, 0x01}},
       // TODO(davidben): Add a change detector for TLS 1.3 once the spec and our
       // implementation has settled enough that it won't change.
   };
@@ -4778,7 +4778,7 @@ TEST(SSLTest, ClientHello) {
       ADD_FAILURE() << "ClientHellos did not match.";
       // Print the value manually so it is easier to update the test vector.
       for (size_t i = 0; i < client_hello.size(); i += 12) {
-        printf("     %c", i == 0 ? '{' : ' ');
+        printf("       %c", i == 0 ? '{' : ' ');
         for (size_t j = i; j < client_hello.size() && j < i + 12; j++) {
           if (j > i) {
             printf(" ");
diff --git a/ssl/test/runner/runner.go b/ssl/test/runner/runner.go
index 3bde782d8..7b58f77fd 100644
--- a/ssl/test/runner/runner.go
+++ b/ssl/test/runner/runner.go
@@ -10169,8 +10169,8 @@ func addSignatureAlgorithmTests() {
 				shouldFail = true
 			}
 
-			// By default, BoringSSL does not enable ecdsa_sha1, ecdsa_secp521_sha512, and ed25519.
-			if alg.id == signatureECDSAWithSHA1 || alg.id == signatureECDSAWithP521AndSHA512 || alg.id == signatureEd25519 {
+			// By default, AWS-LC does not enable ecdsa_sha1 and ed25519.
+			if alg.id == signatureECDSAWithSHA1 || alg.id == signatureEd25519 {
 				rejectByDefault = true
 			}
 
diff --git a/ssl/test/runner/ssl_transfer/test_case_names.txt b/ssl/test/runner/ssl_transfer/test_case_names.txt
index 0ef2c03c8..bee429372 100644
--- a/ssl/test/runner/ssl_transfer/test_case_names.txt
+++ b/ssl/test/runner/ssl_transfer/test_case_names.txt
@@ -639,6 +639,8 @@ Server-VerifyDefault-ECDSA_P256_SHA256-TLS12
 Server-VerifyDefault-ECDSA_P256_SHA256-TLS13
 Server-VerifyDefault-ECDSA_P384_SHA384-TLS12
 Server-VerifyDefault-ECDSA_P384_SHA384-TLS13
+Server-VerifyDefault-ECDSA_P521_SHA512-TLS12
+Server-VerifyDefault-ECDSA_P521_SHA512-TLS13
 Server-VerifyDefault-RSA_PKCS1_SHA1-TLS12
 Server-VerifyDefault-RSA_PKCS1_SHA256-TLS12
 Server-VerifyDefault-RSA_PKCS1_SHA384-TLS12
-- 
2.51.0

